File: Linked Lists/partition.py
def partition(linkedList={}, partition=0):
    '''
    Solution 1 - Stable approach - Two partitions (left/right) not chaging linked list order

    Complexity Analysis
    O(n) time | O(n) space

    Partition a linked list around a value x

    dict: linkedList 
    integer: partition
    return: Linked list re-arranged partition by x    
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False or not isinstance(partition, int) or partition < 0):
        print('Arguments should be a valid non-empty dictionary and a positive valid integer')
        return False

    leftPartition = { 'data': float('inf'), 'next': None }
    rightPartition = { 'data': float('inf'), 'next': None }

    leftCurrNode = leftPartition
    rightCurrNode = rightPartition

    currNode = linkedList

    while (currNode != None):
        if (currNode['data'] < partition):
            leftCurrNode['next']= currNode
            leftCurrNode = leftCurrNode['next']
        else:
            rightCurrNode['next'] = currNode
            rightCurrNode = rightCurrNode['next']

        currNode = currNode['next']

    rightCurrNode['next'] = None
    leftCurrNode['next'] = rightPartition['next']

    return leftPartition['next']

def partition(linkedList={}, partition=0):
    '''
    Solution 2 - Not Stable approach - Rearrange the elements growing the linked list at the head and tail

    Complexity Analysis
    O(n) time | O(1) space

    Partition a linked list around a value x

    dict: linkedList 
    integer: partition
    return: Linked list re-arranged partition by x    
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedList, dict) or bool(linkedList) == False or not isinstance(partition, int) or partition < 0):
        print('Arguments should be a valid non-empty dictionary and a positive valid integer')
        return False

    head = linkedList
    tail = linkedList

    while (linkedList != None):
        linkedListNext = linkedList['next']

        if (linkedList['data'] < partition):
            linkedList['next']= head
            head = linkedList
        else:
            tail['next'] = linkedList
            tail = linkedList

        linkedList = linkedListNext

    tail['next'] = None

    return head

# Test cases (black box - unit testing)
testCases = [
    # Normal
    # Data that is typical (expected) and should be accepted by the system.
    { 
        'assert': partition({ 'data': 5, 'next': { 'data': 2, 'next': { 'data': 9, 'next': { 'data': 0, 'next': None } } } }, 5),
        'expected': { 'data': 2, 'next': { 'data': 0, 'next': { 'data': 5, 'next': { 'data': 9, 'next': None } } } } 
    },
    { 
        'assert': partition({ 'data': 53, 'next': { 'data': 92, 'next': { 'data': 23, 'next': { 'data': 19, 'next': { 'data': 84, 'next': None } } } } }, 23),
        'expected': { 'data': 19, 'next': { 'data': 53, 'next': { 'data': 92, 'next': { 'data': 23, 'next': { 'data': 84, 'next': None } } } } } 
    },
    { 
        'assert': partition({ 'data': 83, 'next': { 'data': 12, 'next': { 'data': 87, 'next': { 'data': 33, 'next': { 'data': 54, 'next': None } } } } }, 87),
        'expected': { 'data': 83, 'next': { 'data': 12, 'next': { 'data': 33, 'next': { 'data': 54, 'next': { 'data': 87, 'next': None } } } } } 
    },

    # Boundary data (extreme data, edge case)
    # Data at the upper or lower limits of expectations that should be accepted by the system.

    # Abnormal data (erroneous data)
    # Data that falls outside of what is acceptable and should be rejected by the system.
    { 'assert': partition(), 'expected': False },
    { 'assert': partition(0), 'expected': False },
    { 'assert': partition(''), 'expected': False },
    { 'assert': partition([]), 'expected': False },
    { 'assert': partition(()), 'expected': False },
    { 'assert': partition({}), 'expected': False },
    { 'assert': partition(None), 'expected': False },
    { 'assert': partition(False), 'expected': False }
]

# Run tests
for (index, test) in enumerate(testCases):
    print(f'# Test {index + 1}')
    print(f'Actual: {test["assert"]}')
    print(f'expected: {test["expected"]}')
    print('🤘 Test PASSED 🤘' if test["assert"] == test["expected"] else '👎 Test FAILED 👎', '\n')

File: Linked Lists/partition.js
import { factoryLinkedListNode as createNode } from '../../Data Structures/linkedList.js';

/**
 * Solution 1 - Stable approach - Two partitions (left/right) not chaging linked list order
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Partition a linked list around a value x
 *
 * @param {object} linkedList
 * @param {number} partition
 * @returns {object} Linked list re-arranged partition by x
 */
function partition(linkedList = {}, partition = 0) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0 || typeof partition !== 'number' || partition < 0) {
        console.error('Parameters must be a valid non-empty object and a positive valid number');
        return false;
    }

    const leftPartition = createNode();
    const rightPartition = createNode();

    let leftCurrNode = leftPartition;
    let rightCurrNode = rightPartition;

    let currNode = linkedList;

    while (currNode !== null) {
        if (currNode.data < partition) {
            leftCurrNode.next = currNode;
            leftCurrNode = leftCurrNode.next;
        } else {
            rightCurrNode.next = currNode;
            rightCurrNode = rightCurrNode.next;
        }

        currNode = currNode.next;
    }

    rightCurrNode.next = null;
    leftCurrNode.next = rightPartition.next;

    return leftPartition.next;
}

/**
 * Solution 2 - Not Stable approach - Rearrange the elements growing the linked list at the head and tail
 *
 * Complexity Analysis
 * O(n) time | O(1) space
 *
 * Partition a linked list around a value x
 *
 * @param {object} linkedList
 * @param {number} partition
 * @returns {object} Linked list re-arranged partition by x
 */
function partition(linkedList = {}, partition = 0) {
    // Gracefully handle type and Falsy values
    if (typeof linkedList !== 'object' || Object.keys(linkedList || {}).length === 0 || typeof partition !== 'number' || partition < 0) {
        console.error('Parameters must be a valid non-empty object and a positive valid number');
        return false;
    }

    let head = linkedList;
    let tail = linkedList;

    while (linkedList !== null) {
        let next = linkedList.next;

        if (linkedList.data < partition) {
            linkedList.next = head;
            head = linkedList;
        } else {
            tail.next = linkedList;
            tail = linkedList;
        }

        linkedList = next;
    }

    tail.next = null;

    return head;
}

const testCases = [
    // Normal
    // Data that is typical (expected) and should be accepted by the system.
    { 
        assert: JSON.stringify(partition({ data: 5, next: { data: 2, next: { data: 10, next: null } } }, 5)),
        expected: JSON.stringify({ data: 2, next: { data: 5, next: { data: 10, next: null } } }),
    },
    { 
        assert: JSON.stringify(partition({ data: 1, next: { data: 5, next: { data: 4, next: { data: 3, next: { data: 2, next: null } } } } }, 3)),
        expected: JSON.stringify({ data: 1, next: { data: 2, next: { data: 5, next: { data: 4, next: { data: 3, next: null } } } } }),
    },
    { 
        assert: JSON.stringify(partition({ data: 9, next: { data: 2, next: { data: 6, next: { data: 4, next: { data: 3, next: null } } } } }, 4)),
        expected: JSON.stringify({ data: 2, next: { data: 3, next: { data: 9, next: { data: 6, next: { data: 4, next: null } } } } }),
    },
    { 
        assert: JSON.stringify(partition({ data: 14, next: { data: 2, next: { data: 8, next: { data: 6, next: { data: 3, next: { data: 1, next: { data: 11, next: null } } } } } } }, 8)),
        expected: JSON.stringify({ data: 2, next: { data: 6, next: { data: 3, next: { data: 1, next: { data: 14, next: { data: 8, next: { data: 11, next: null } } } } } } }),
    },

    // Boundary data (extreme data, edge case)
    // Data at the upper or lower limits of expectations that should be accepted by the system.
    { 
        assert: JSON.stringify(partition({ data: 1, next: null }, 2)),
        expected: JSON.stringify({ data: 1, next: null }),
    },
    { 
        assert: JSON.stringify(partition({ data: 10, next: { data: 100, next: null } }, 34)),
        expected: JSON.stringify({ data: 10, next: { data: 100, next: null } }),
    },

    // Abnormal data (erroneous data)
    // Data that falls outside of what is acceptable and should be rejected by the system.
    { assert: partition(), expected: false },
    { assert: partition(0), expected: false },
    { assert: partition(''), expected: false },
    { assert: partition([]), expected: false },
    { assert: partition({}), expected: false },
    { assert: partition(NaN), expected: false },
    { assert: partition(null), expected: false },
    { assert: partition(false), expected: false },
    { assert: partition(new Set()), expected: false },
    { assert: partition(new Map()), expected: false },
    { assert: partition(undefined), expected: false },
];

// Run tests
testCases.forEach((test, index) => {
    const currentTest = `# Test ${index + 1}`;

    console.group(currentTest);
    console.log('Actual:', test.assert);
    console.log('Expected:', test.expected);
    console.log(test.assert === test.expected ? '🤘 Test PASSED 🤘' : '👎 Test FAILED 👎', '\n');
    console.groupEnd(currentTest);
});
